pkg/beacon/dkg/dkg.go

```diff
@@ -131,44 +133,56 @@ func decideMemberFate(
 	startPublicationBlockHeight uint64,
 	beaconChain beaconchain.Interface,
 	blockCounter chain.BlockCounter,
-) error {
+) ([]group.MemberIndex, error) {
 	dkgResultEvent, err := waitForDkgResultEvent(
 		dkgResultChannel,
 		startPublicationBlockHeight,
 		beaconChain,
 		blockCounter,
 	)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	groupPublicKey, err := gjkrResult.GroupPublicKeyBytes()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	// If member don't support the same group public key, it could not stay
 	// in the group.
 	if !bytes.Equal(groupPublicKey, dkgResultEvent.GroupPublicKey) {
-		return fmt.Errorf(
+		return nil, fmt.Errorf(
 			"[member:%v] could not stay in the group because "+
 				"member do not support the same group public key",
 			playerIndex,
 		)
 	}
 
+	misbehavedSet := make(map[group.MemberIndex]struct{})
+	for _, misbehavedID := range dkgResultEvent.Misbehaved {
+		misbehavedSet[misbehavedID] = struct{}{}
+	}
+
 	// If member is considered as misbehaved, it could not stay in the group.
-	for _, misbehaved := range dkgResultEvent.Misbehaved {
-		if playerIndex == misbehaved {
-			return fmt.Errorf(
-				"[member:%v] could not stay in the group because "+
-					"member is considered as misbehaving",
-				playerIndex,
-			)
+	if _, isMisbehaved := misbehavedSet[playerIndex]; isMisbehaved {
+		return nil, fmt.Errorf(
+			"[member:%v] could not stay in the group because "+
+				"member is considered as misbehaving",
+			playerIndex,
+		)
+	}
+
+	// Construct a new view of the operating members according to the accepted
+	// DKG result.
+	operatingMemberIDs := make([]group.MemberIndex, 0)
+	for _, memberID := range gjkrResult.Group.MemberIDs() {
+		if _, isMisbehaved := misbehavedSet[memberID]; !isMisbehaved {
+			operatingMemberIDs = append(operatingMemberIDs, memberID)
```

Comment:
But `operatingMemberIDs` just hold indexes of members who are still operating. The order does not matter. This is exactly how `Group.OperatingMemberIDs` work:
```
func (g *Group) OperatingMemberIDs() []MemberIndex {
	operatingMembers := make([]MemberIndex, 0)
	for _, member := range g.MemberIDs() {
		if g.IsOperating(member) {
			operatingMembers = append(operatingMembers, member)
		}
	}

	return operatingMembers
}
```

Reply:
You are right, this function is fine BUT I think this problem exist in `resolveGroupOperators`. The produced `groupMembers` array is shorter than the group size. This is saved in `ThresholdSigner` that ends up with `groupPublicKeyShares` not maching member indices in `groupOperators`.

Reply:
Wait, no, you are right and this is correct:
1. On-chain it is expected that the members hash contain only members who actively participated in DKG:
https://github.com/keep-network/keep-core/blob/dc6d086a3a494e2aa3cd56e02d8f2bd3ae69f219/solidity/random-beacon/contracts/libraries/BeaconDkg.sol#L86-L88
https://github.com/keep-network/keep-core/blob/dc6d086a3a494e2aa3cd56e02d8f2bd3ae69f219/solidity/random-beacon/contracts/BeaconDkgValidator.sol#L260-L286

2. Off-chain we let to declare member ID in `SignatureShareMessage` because we validate the key share anyway: 
https://github.com/keep-network/keep-core/blob/dc6d086a3a494e2aa3cd56e02d8f2bd3ae69f219/pkg/beacon/entry/entry.go#L193